### PR TITLE
Remove `AssignInst`.

### PR DESCRIPTION
**Pull request by ltratt**
This PR removes `AssignInst` from the JIT IR: in essence we used that to simplify trace_builder, but that's not a great trade-off, because it complicates every other part of the system.

The two main commits here are to remove `AssignInst` from phi nodes (https://github.com/ykjit/yk/commit/2227b67f05a5dab7f39f4df90bf09166e541425d) and pointer adds (https://github.com/ykjit/yk/commit/45e9b77bee65d4c62c0eac1ed75841cdad619b3c). In particular the phi node commit is somewhat subtle, and the commit message hopefully explains what it's doing.

### CODE REVIEW

File: ykrt/src/compile/jitc_yk/trace_builder.rs [hunk 1]
```diff
@@ -895,8 +895,14 @@ impl<'a> TraceBuilder<'a> {
...
         let chosen_val = &incoming_vals[incoming_bbs.iter().position(|bb| bb == prev_bb).unwrap()];
-        let assign = jit_ir::AssignInst::new(&self.handle_operand(chosen_val)?);
-        self.copy_inst(assign.into(), bid, aot_inst_idx)
+        let aot_iit = aot_ir::InstID::new(
+            bid.func_idx(),
+            bid.bb_idx(),
+            aot_ir::InstIdx::new(aot_inst_idx),
+        );
+        let op = self.handle_operand(chosen_val)?;
+        self.local_map.insert(aot_iit, op);
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/trace_builder.rs:911]: Good, this is what I was looking for in this commit. This is how we did it in the LLVM backend also with the VMap.

File: tests/c/ptradd.newcg.c [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+//     %{{18_5}}: ptr = ptr_add %{{18_3}}, 0 + (%{{18_4}} * 1)
+//     ...
+//     --- End aot ---
+//     --- Begin jit-pre-opt ---
+//     ...
+//     %{{21}}: i32 = call @f()
+//     ...
+//     %{{25}}: i32 = add %{{21}}, 1i32
+//     ...
+//     %{{29}}: ptr = dyn_ptr_add %{{26}}, %{{27}}, 1
```
[the remark this answers is not in the document]
Comment (ltratt) [reply tests/c/ptradd.newcg.c:25]: There are tons of intermediate instructions, unfortunately, including things like shadowstack stuff, and using C input for tests means that it's difficult to really test exactly what we want to test.

[the remark this answers is not in the document]
Comment (ptersilie) [reply tests/c/ptradd.newcg.c:25]: Ah. fair enough.

File: tests/c/ptradd.newcg.c [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+//     %{{_}}: i32 = call @fprintf(%{{_}}, %{{_}}, %{{_}}, %{{_}}, %{{_}})
+//     ...
+//     --- End jit-pre-opt ---
+//     i=3, val=3, p=8
+//     jitstate: enter-jit-code
+//     i=2, val=3, p=12
+//     i=1, val=3, p=16
+//     jitstate: deoptimise
+
+// Check that PHI nodes JIT properly.
```
[the remark this answers is not in the document]
Comment (ltratt) [reply tests/c/ptradd.newcg.c]: Oops. You might be able to tell that I copied a test over. Fixed in rough.

File: ykrt/src/compile/jitc_yk/trace_builder.rs [hunk 4]
```diff
@@ -299,25 +299,34 @@ impl<'a> TraceBuilder<'a> {
         Ok(())
     }
 
+    fn link_iid_to_last_instr(&mut self, bid: &aot_ir::BBlockId, aot_inst_idx: usize) {
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/trace_builder.rs:302]: A docstring may help this function. I was curious as to what it did.

File: tests/c/ptradd.newcg.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ltratt) [reply tests/c/ptradd.newcg.c]: > There are existing newcg C tests for both ... ptr_adds.

None of the existing tests were triggered by this test, so I'm not sure which are really testing what they should be... Happy to be corrected!

File: ykrt/src/compile/jitc_yk/trace_builder.rs (hunk 4 above)
Comment (ptersilie) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:302]: I was going to ask the same.

File: ykrt/src/compile/jitc_yk/trace_builder.rs [hunk 5]
```diff
@@ -751,10 +760,8 @@ impl<'a> TraceBuilder<'a> {
                 jit_ir::DynPtrAddInst::new(jit_ptr, num_elems, elem_size).into(),
             )?;
         }
-
-        // OPT: the assignment instruction could be elided in some cases.
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/trace_builder.rs:755]: Yeah good. I knew this was a hack when I wrote it :)

File: ykrt/src/compile/jitc_yk/trace_builder.rs (hunk 4 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:303]: I probably should, but this part of trace_builder is a mess, and I hope to somewhat tidy it up soon. Still, fixed in the interim in ea1748d9.

File: tests/c/ptradd.newcg.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (vext01) [reply tests/c/ptradd.newcg.c]: It's quite similar to https://github.com/ykjit/yk/blob/master/tests/c/neg_ptradd_dyn_ptr.newcg.c

I'm happy to leave the new test in, maybe with a more descriptive name?

I note that the existing test names aren't all that uniform, but maybe not worth worrying about?

```
tests/c/dyn_ptradd_mixed.newcg.c     tests/c/neg_ptradd_dyn.newcg.c
tests/c/dyn_ptradd_multidim.newcg.c  tests/c/neg_ptradd_dyn_ptr.newcg.c
tests/c/dyn_ptradd_simple.newcg.c    tests/c/neg_ptradd.newcg.c
```

[the remark this answers is not in the document]
Comment (ltratt) [reply tests/c/ptradd.newcg.c]: I hope that these tests will one day become unit tests: I don't know that worrying about the names now is useful TBH.

File: ykrt/src/compile/jitc_yk/trace_builder.rs (hunk 4 above)
Comment (vext01) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:303]: Don't worry about this if your next PR will delete this function, but:

> Link the AOT IR to the last instruction pushed into the JIT IR.

The specified AOT IR *BBlockId*, right?

Comment (vext01) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:303]: Ah never mind.

**Comment by vext01** [comment]
Please squash.

**Comment by ltratt** [comment]
Squashed.

**Pull request merged**